### PR TITLE
Server fuzzer for binder transport

### PR DESCRIPTION
**Pull request by TaWeiTu**

### CODE REVIEW

File: src/core/ext/transport/binder/utils/transport_stream_receiver_impl.cc [hunk 1]
```diff
@@ -117,7 +117,7 @@ void TransportStreamReceiverImpl::RegisterRecvTrailingMetadata(
 void TransportStreamReceiverImpl::NotifyRecvInitialMetadata(
     StreamIdentifier id, absl::StatusOr<Metadata> initial_metadata) {
   gpr_log(GPR_INFO, "%s id = %d is_client = %d", __func__, id, is_client_);
-  if (!is_client_ && accept_stream_callback_) {
+  if (!is_client_ && accept_stream_callback_ && initial_metadata.ok()) {
```
Comment (sifmelcara) [inline src/core/ext/transport/binder/utils/transport_stream_receiver_impl.cc:120]: Why this change in this PR?

File: test/core/transport/binder/end2end/fuzzers/client_fuzzer.cc [hunk 2]
```diff
@@ -12,228 +12,24 @@
...
-BinderForFuzzing::ConstructTxReceiver(
-    grpc_core::RefCountedPtr<grpc_binder::WireReader> wire_reader_ref,
-    grpc_binder::TransactionReceiver::OnTransactCb cb) const {
-  return absl::make_unique<TransactionReceiverForFuzzing>(data_, size_,
-                                                          wire_reader_ref, cb);
-}
-
-void* tag(intptr_t t) { return reinterpret_cast<void*>(t); }
-void dont_log(gpr_log_func_args*) {}
+static void* tag(intptr_t t) { return reinterpret_cast<void*>(t); }
```
Comment (sifmelcara) [inline test/core/transport/binder/end2end/fuzzers/client_fuzzer.cc:30]: Why not anonymous namespace

File: test/core/transport/binder/end2end/fuzzers/fuzzer_utils.cc [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "test/core/transport/binder/end2end/fuzzers/fuzzer_utils.h"
+
+namespace grpc_binder {
+namespace fuzzing {
+
+std::thread* g_fuzzing_thread = nullptr;
```
Comment (sifmelcara) [inline test/core/transport/binder/end2end/fuzzers/fuzzer_utils.cc:20]: Can you put this variable in anonymous namespace and provide `CreateFuzzingThread()` `GetFuzzingThread()` `DeleteFuzzingThread()` helper functions to access it?

(It is a bit messy to access global variable defined in other files directly)

File: test/core/transport/binder/end2end/fuzzers/server_fuzzer.cc [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+#include "src/core/ext/transport/binder/transport/binder_transport.h"
+#include "src/core/lib/iomgr/executor.h"
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/lib/surface/server.h"
+#include "test/core/transport/binder/end2end/fuzzers/fuzzer_utils.h"
+
+bool squelch = true;
+bool leak_check = true;
+
+static void* tag(intptr_t t) { return reinterpret_cast<void*>(t); }
```
Comment (sifmelcara) [inline test/core/transport/binder/end2end/fuzzers/server_fuzzer.cc:26]: Use anonymous namespace?

File: test/core/transport/binder/end2end/fuzzers/fuzzer_utils.cc (hunk 3 above)
Comment (sifmelcara) [reply test/core/transport/binder/end2end/fuzzers/fuzzer_utils.cc:20]: Or, alternatively, define `g_fuzzing_thread` in client fuzzer / server fuzzer `.cc` file and let caller pass it into `TranasctionReceiverForFuzzing::TranasctionReceiverForFuzzing()`

File: test/cpp/microbenchmarks/bm_binder/java/io/grpc/binder/cpp/benchmarks/server/GrpcService.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
+package io.grpc.binder.cpp.benchmarks.server;
```
Comment (sifmelcara) [inline test/cpp/microbenchmarks/bm_binder/java/io/grpc/binder/cpp/benchmarks/server/GrpcService.java:1]: Accidentally committed?

File: test/core/transport/binder/end2end/fuzzers/client_fuzzer.cc (hunk 2 above)
Comment (TaWeiTu) [reply test/core/transport/binder/end2end/fuzzers/client_fuzzer.cc:30]: These are copied from HTTP/2 fuzzers and I feel like it's not necessarily good to change them? It was probably a mistake to put them in anonymous namespace before.

File: test/core/transport/binder/end2end/fuzzers/server_fuzzer.cc (hunk 4 above)
Comment (TaWeiTu) [reply test/core/transport/binder/end2end/fuzzers/server_fuzzer.cc:26]: Same as above. I want to keep everything copied clean so that the difference stands out.

File: test/cpp/microbenchmarks/bm_binder/java/io/grpc/binder/cpp/benchmarks/server/GrpcService.java (hunk 5 above)
Comment (TaWeiTu) [reply test/cpp/microbenchmarks/bm_binder/java/io/grpc/binder/cpp/benchmarks/server/GrpcService.java]: Oops. Fixed

File: test/core/transport/binder/end2end/fuzzers/fuzzer_utils.cc (hunk 3 above)
Comment (TaWeiTu) [reply test/core/transport/binder/end2end/fuzzers/fuzzer_utils.cc:22]: Done.

File: src/core/ext/transport/binder/utils/transport_stream_receiver_impl.cc (hunk 1 above)
Comment (TaWeiTu) [reply src/core/ext/transport/binder/utils/transport_stream_receiver_impl.cc:120]: We probably don't want to open a new stream for invalid initial metadata. This was inspired by the server fuzzer and since this is so small I included it here. I can also open a separate PR for this.

File: test/core/transport/binder/end2end/fuzzers/client_fuzzer.cc (hunk 2 above)
Comment (sifmelcara) [reply test/core/transport/binder/end2end/fuzzers/client_fuzzer.cc:30]: Hmm I think our team generally prefers anonymous namespace more but it is ok if HTTP/2 fuzzers are using static

File: test/core/transport/binder/end2end/fuzzers/fuzzer_utils.cc [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+template <typename... Args>
+void CreateFuzzingThread(Args&&... args) {
+  GPR_ASSERT(g_fuzzing_thread == nullptr);
+  g_fuzzing_thread = new std::thread(std::forward<Args>(args)...);
+}
+
+}  // namespace
+
+void JoinFuzzingThread() {
```
Comment (sifmelcara) [inline test/core/transport/binder/end2end/fuzzers/fuzzer_utils.cc:32]: Rename to something like join and delete?

**Review by sifmelcara — APPROVED** [review]
Still need rebase

File: test/core/transport/binder/end2end/fuzzers/client_fuzzer.cc (hunk 2 above)
Comment (TaWeiTu) [reply test/core/transport/binder/end2end/fuzzers/client_fuzzer.cc:30]: Yeah I guess that's because it was written before transitioning to C++.

**Comment by sifmelcara** [comment]
Rebase and rerun CI

**Comment by sifmelcara** [comment]
cc @napathome
